### PR TITLE
Sparse updates for distributed learning

### PR DESCRIPTION
**Pull request by MissingNO57**

### CODE REVIEW

File: libs/dmlf/include/dmlf/collective_learning/translator.hpp [hunk 1]
```diff
@@ -47,55 +87,31 @@ struct Translator
...
       mask.Fill(DataType{1});
-      return std::make_pair(gradient_update, mask);
+      return std::make_pair(weights, mask);
     }
 
     assert(VocabKnown(vocab_hash));
 
     std::vector<std::string> other_vocab = known_vocabs[vocab_hash];
 
     // figure out which way around the matrix is
```
Comment (evsmithx) [inline libs/dmlf/include/dmlf/collective_learning/translator.hpp]: Outdated comment

File: libs/ml/tests/unit/utilities/sparse.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  TypeParam data = utilities::FromSparse(sparse_data, rows, 5);
+
+  std::vector<SizeType> gt_shape = {3, 5};
+  EXPECT_EQ(data.shape(), gt_shape);
+
+  EXPECT_TRUE(data.View(4).Copy().AllClose(sparse_data.View(0).Copy()));
+  EXPECT_TRUE(data.View(3).Copy().AllClose(TypeParam({3, 1})));
+  EXPECT_TRUE(data.View(2).Copy().AllClose(sparse_data.View(1).Copy()));
+  EXPECT_TRUE(data.View(1).Copy().AllClose(sparse_data.View(2).Copy()));
+  EXPECT_TRUE(data.View(0).Copy().AllClose(TypeParam({3, 1})));
```
Comment (evsmithx) [inline libs/ml/tests/unit/utilities/sparse.cpp:57]: No hard-coded random numbers, yay!

**Review by evsmithx — DISMISSED** [review]
One outdated comment, but the rest of it looks good!

File: libs/ml/include/ml/core/graph.hpp [hunk 3]
```diff
@@ -938,6 +940,15 @@ std::vector<TensorType> Graph<TensorType>::GetGradientsReferences() const
   return ret;
 }
 
+template <typename TensorType>
```
Comment (baykaner) [inline libs/ml/include/ml/core/graph.hpp:950]: function comment please. Please describe why this function exists and when we might invoke it.
it's not obvious why you get a vector of unordered sets here unless you already have the word2vec application in mind.

**Review by baykaner — CHANGES_REQUESTED** [review]

**Review by baykaner — APPROVED** [review]

**Pull request merged**